### PR TITLE
Fixed #26348 -- Added TruncTime and exposed it through the __time lookup.

### PR DESCRIPTION
**Pull request by charettes**
https://code.djangoproject.com/ticket/26348

### CODE REVIEW

**Comment by charettes** [comment]
@jarshwah, I'd like to have your input on this when you're available.

File: docs/ref/models/database-functions.txt [hunk 1]
```diff
@@ -286,10 +286,12 @@ Date Functions
 We'll be using the following model in examples of each function::
 
     class Experiment(models.Model):
-        start_time = models.DateTimeField()
+        start_datetime = models.DateTimeField()
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: It would be nice not to rename all these fields, but if we must do it, let's make the changes in a separate commit so we can backport to 1.10.

File: docs/ref/models/database-functions.txt [hunk 2]
```diff
@@ -616,6 +619,61 @@ that deal with date-parts can be used with ``DateField``::
     2016-01-01 00:00:00+11:00 1
     2014-06-01 00:00:00+10:00 1
 
+.. versionadded:: 1.11
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: These annotations go below the headings.

File: docs/ref/models/database-functions.txt [hunk 3]
```diff
@@ -628,6 +686,18 @@ that deal with date-parts can be used with ``DateField``::
 truncate function. It's also registered as a transform on  ``DateTimeField`` as
 ``__date``.
 
+.. versionadded:: 1.11
+
+.. class:: TruncTime(expression, **extra)
+
+    .. attribute:: lookup_name = 'time'
+    .. attribute:: output_field = TimeField()
+
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: chop a blank line

**Comment by jarshwah** [comment]
Just letting you know I do have this in my queue, but just don't have time right now. Will try and have a look during the week.

What were your plans regarding Oracle? Were you looking for advice on how to tackle it, or help from someone with access to an Oracle database?

**Comment by charettes** [comment]
No worries Josh, it's targeting 1.11 as a new feature anyway I just wanted to get familar with the internals of `Trunc` and friends.

I should be able to add Oracle support in the next few days, I've got the VM setup locally. I just didn't feel like firing it up before getting any input from your side.

**Comment by jarshwah** [comment]
Without going into a lot of detail it looks reasonably good to me. The only concern is really with something that I began - "not a DateField or DateTimeField or TimeField" seems messy from a user perspective, even though it simplifies the code. If there's a nice way to avoid this kind of error without duplicating code everywhere I think that'd be worthwhile.

Will give it a better review this week sometime.

File: docs/ref/models/database-functions.txt (hunk 1 above)
Comment (charettes) [inline docs/ref/models/database-functions.txt]: Done.

File: docs/ref/models/database-functions.txt (hunk 3 above)
Comment (charettes) [inline docs/ref/models/database-functions.txt]: Done.

**Comment by charettes** [comment]
Buildbot, test on oracle.

File: docs/ref/models/database-functions.txt [hunk 4]
```diff
@@ -616,6 +619,61 @@ that deal with date-parts can be used with ``DateField``::
...
+.. class:: TruncSecond(expression, output_field=None, tzinfo=None, **extra)
+
+    .. attribute:: kind = 'second'
+
+These are logically equivalent to ``Trunc('time_field', kind)``. They truncate
+all parts of the time up to ``kind`` which allows grouping or filtering times
+with less precision. ``expression`` can have an ``output_field`` of either
+``TimeField`` or ``DateTimeField``.
+
+Since ``TimeField``\s don't have a time component, only ``Trunc`` subclasses
```
Comment (jarshwah) [inline docs/ref/models/database-functions.txt]: don't have a *date* component

**Comment by charettes** [comment]
@jarshwah I'm thinking of fixing the "not a DateField or DateTimeField or TimeField" case in another commit.

**Comment by charettes** [comment]
The only remaining failure is related to `django.db.backends.oracle.operations.DatabaseOperations._convert_field_to_tz` stripping sub-seconds by casting to `DATE()`.

@jarshwah, @shaib any idea how this could be solved?

**Comment by charettes** [comment]
77b73e79a4750dcbfabc528bf00cad81ff5bb4d9 looks like the culprit.

**Comment by charettes** [comment]
Hmm it looks like Oracle timezone conversion always dropped sub-second precision (https://github.com/django/django/commit/e74e207cce54802f897adcb42149440ee154821e#diff-54b46d05e1da568b3cc987c423e00c50R166) and https://github.com/django/django/commit/77b73e79a4750dcbfabc528bf00cad81ff5bb4d9 was just an adjustment to make it behave consistently when `USE_TZ=False`.

**Comment by charettes** [comment]
Buildbot, test on oracle.

**Comment by jarshwah** [comment]
Yes, I was surprised by the behaviour of the timezone casting too. I think I was meant to come back to it and forgot to do so. Nice to see you've fixed it up, thanks!

**Comment by timgraham** [comment]
Add a mention of the new DatabaseOperations method in the release notes, similar to `DatabaseOperations.time_extract_sql` in 1.9.

File: django/db/backends/oracle/operations.py [hunk 5]
```diff
@@ -128,6 +128,13 @@ def datetime_cast_date_sql(self, field_name, tzname):
         sql = 'TRUNC(%s)' % field_name
         return sql, []
 
+    def datetime_cast_time_sql(self, field_name, tzname):
+        # Since `TimeField` are stored as TIMESTAMP where the "date" part
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: values are stored

File: django/db/backends/oracle/operations.py [hunk 6]
```diff
@@ -128,6 +128,13 @@ def datetime_cast_date_sql(self, field_name, tzname):
...
+        # is ignored only make sure to convert the field to the specified
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: I think this means: "only the date part is ignored,"
drop "make sure to"

File: docs/ref/models/database-functions.txt [hunk 7]
```diff
@@ -686,6 +686,17 @@ that deal with time-parts can be used with ``TimeField``::
 truncate function. It's also registered as a transform on  ``DateTimeField`` as
 ``__date``.
 
+.. versionadded:: 1.11
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: this goes below the `.. class::`

File: docs/ref/models/querysets.txt [hunk 8]
```diff
@@ -2674,6 +2674,27 @@ When :setting:`USE_TZ` is ``True``, datetime fields are converted to the
 current time zone before filtering. This requires :ref:`time zone definitions
 in the database <database-time-zone-definitions>`.
 
+.. fieldlookup:: time
+
+.. versionadded:: 1.11
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: put below heading

File: docs/ref/models/querysets.txt [hunk 9]
```diff
@@ -2674,6 +2674,27 @@ When :setting:`USE_TZ` is ``True``, datetime fields are converted to the
...
+
+``time``
+~~~~~~~~
+
+For datetime fields, casts the value as time. Allows chaining additional field
+lookups. Takes a time value.
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: :class:\`datetime.time\` and/or time.time (use a link to clarify)?

File: docs/releases/1.11.txt [hunk 10]
```diff
@@ -194,6 +194,8 @@ Models
 
 * :class:`~django.db.models.ImageField` now has a default
   :data:`~django.core.validators.validate_image_file_extension` validator.
+* Added support for time truncation to
```
Comment (timgraham) [inline docs/releases/1.11.txt]: add blank line above

**Comment by timgraham** [comment]
Repeated from last review: Add a mention of the new DatabaseOperations methods in the release notes, similar to DatabaseOperations.time_extract_sql in 1.9.

**Comment by charettes** [comment]
> Repeated from last review: Add a mention of the new DatabaseOperations methods in the release notes, similar to DatabaseOperations.time_extract_sql in 1.9.

Will do, sorry I missed that.

**Comment by charettes** [comment]
@timgraham, just pushed the missing release notes.

File: docs/releases/1.11.txt [hunk 11]
```diff
@@ -263,7 +266,12 @@ Backwards incompatible changes in 1.11
 Database backend API
 --------------------
 
-* ...
+* The ``DatabaseOperations.time_trunc_sql()`` method was added to support
```
Comment (timgraham) [inline docs/releases/1.11.txt]: I've been using "is" instead of "was".

File: docs/releases/1.11.txt [hunk 12]
```diff
@@ -263,7 +266,12 @@ Backwards incompatible changes in 1.11
...
+  ``TimeField`` truncation. It should accept a ``lookup_type`` and
```
Comment (timgraham) [inline docs/releases/1.11.txt]: It accepts... and returns...

File: docs/releases/1.11.txt [hunk 13]
```diff
@@ -273,6 +277,11 @@ Database backend API
   specificity. The ``lookup_type`` argument can be either ``'hour'``,
   ``'minute'``, or ``'second'``.
 
+* The ``DatabaseOperations.datetime_cast_time_sql()`` method was added to
```
Comment (timgraham) [inline docs/releases/1.11.txt]: Same comments as in other commit.
